### PR TITLE
Add instrumental forest tuning

### PR DESCRIPTION
**Pull request by erikcs**
Towards #501 (core instrumental forest tuning functionality is in place with this PR):

* [C++: InstrumentalPredictionStrategy::compute_error](https://github.com/erikcs/grf/blob/IV-tuning/core/src/prediction/InstrumentalPredictionStrategy.cpp#L206) (`debiased.error` used for tuning) is redefined to compute the reduced form error for iv forest. For causal_forest (instrumental_forest with Z=W) the output is unchanged.

* R:  Add [tune_instrumental_forest](https://github.com/erikcs/grf/blob/IV-tuning/r-package/grf/R/tune_instrumental_forest.R) 

    Add accompanying [test](https://github.com/erikcs/grf/blob/IV-tuning/r-package/grf/tests/testthat/test_instrumental_forest_tuning.R) that tuning reduces MSE

Remaining issues to be addressed in a follow up PR: 

1. Settle on a potentially new parameter grid for IV forest
2. Decide which parameters to tune by default

### CODE REVIEW

**Comment by erikcs** [comment]
An example from the DGP in #501 @ baff582

```R
n = 3000; p = 5
reps = replicate(25, {
  X = matrix(rbinom(n*p, 1, 0.5), n, p)
  Z = rbinom(n, 1, 0.5)
  Q = rbinom(n, 1, 0.5)
  T = Q * Z
  eps = rnorm(n)
  TAU = X[,1] / 2
  Y = rowSums(X[,1:3]) + TAU * T + Q + eps

  # ignore y.hat, w.hat, z.hat tuning
  Y.hat = predict(regression_forest(X, Y, num.trees = 500))$predictions
  W.hat = predict(regression_forest(X, T, num.trees = 500))$predictions
  Z.hat = predict(regression_forest(X, Z, num.trees = 500))$predictions

  iv1 = instrumental_forest(X, Y, T, Z, Y.hat, W.hat, Z.hat)
  iv2 = instrumental_forest(X, Y, T, Z, Y.hat, W.hat, Z.hat, min.node.size = 50)
  iv3 = instrumental_forest(X, Y, T, Z, Y.hat, W.hat, Z.hat, tune.parameters = "all")
  iv4 = instrumental_forest(X, Y, T, Z, Y.hat, W.hat, Z.hat, tune.parameters = "min.node.size")

  c(mean((TAU - predict(iv1)$predictions)^2),
    mean((TAU - predict(iv2)$predictions)^2),
    mean((TAU - predict(iv3)$predictions)^2),
    mean((TAU - predict(iv4)$predictions)^2)
  )
})
rowMeans(reps)
[1] 0.17008953 0.02888249 0.05312565 0.03261020
```

Setting `min.node.size = 50` beats all, even tuning only `min.node.size`. If I naively change the min.node.size [grid](https://github.com/grf-labs/grf/blob/master/r-package/grf/R/tuning.R#L142) to U~(25, 125) it finds the better one with mse 0.0261066. This does not give a big improvement with tune.parameters = "all", suggesting the tuning grids could need some calibration. (increasing `tune.num.reps` and `tune.num.draws` did not help)

File: core/src/prediction/InstrumentalPredictionStrategy.cpp [hunk 1]
```diff
@@ -206,15 +210,15 @@ std::vector<std::pair<double, double>> InstrumentalPredictionStrategy::compute_e
     const Data& data) const {
 
   double instrument_effect_numerator = average.at(OUTCOME_INSTRUMENT) - average.at(OUTCOME) * average.at(INSTRUMENT);
```
Comment (swager) [inline core/src/prediction/InstrumentalPredictionStrategy.cpp]: For consistency, we should probably call this `reduced_form_numerator`

File: core/src/prediction/InstrumentalPredictionStrategy.cpp [hunk 2]
```diff
@@ -240,14 +244,15 @@ std::vector<std::pair<double, double>> InstrumentalPredictionStrategy::compute_e
     }
     const std::vector<double>& leaf_value = leaf_values.get_values(n);
     double outcome_loto = (num_trees *  average.at(OUTCOME) - leaf_value.at(OUTCOME)) / (num_trees - 1);
-    double treatment_loto = (num_trees *  average.at(TREATMENT) - leaf_value.at(TREATMENT)) / (num_trees - 1);
     double instrument_loto = (num_trees *  average.at(INSTRUMENT) - leaf_value.at(INSTRUMENT)) / (num_trees - 1);
     double outcome_instrument_loto = (num_trees *  average.at(OUTCOME_INSTRUMENT) - leaf_value.at(OUTCOME_INSTRUMENT)) / (num_trees - 1);
-    double treatment_instrument_loto = (num_trees *  average.at(TREATMENT_INSTRUMENT) - leaf_value.at(TREATMENT_INSTRUMENT)) / (num_trees - 1);
+    double instrument_instrument_loto = (num_trees *  average.at(INSTRUMENT_INSTRUMENT) - leaf_value.at(INSTRUMENT_INSTRUMENT)) / (num_trees - 1);
+
     double instrument_effect_numerator_loto = outcome_instrument_loto - outcome_loto * instrument_loto;
```
Comment (swager) [inline core/src/prediction/InstrumentalPredictionStrategy.cpp]: same here

**Comment by erikcs** [comment]
For the [above](https://github.com/grf-labs/grf/pull/547#issuecomment-546570635) DGP 

1. changing the `min.node.size` grid to produce larger nodes gives an improvement in MSE but it can have a detrimental effect in other DGPs.

2. Changing which parameters to tune can greatly affect the tuned MSE. For example, tuning `all` might be worse than just tuning `min.node.size`.

@swager Shall we merge this now and handle 1. and 2. in a future PR?

**Pull request merged**